### ISSUE TITLE
VimwikiRenameLink ruins window layout

### ISSUE THREAD

**Issue by ghost** [vimwiki/vimwiki#592]
I run vim 8 with patches 1-1519 and the dev branch of vimwiki installed via vim-plug like this:

    Plug 'vimwiki/vimwiki', { 'branch' : 'dev' }

I have four open tabs. First two tabs are outlines managed by https://github.com/vim-voom/VOoM. Last two tabs are two different vimwikis.

I go to the third tab, enter a vimwiki page and run `:VimwikiRenameLink` . After it finishes working, the layout is totally destroyed:

* tabs 3 and 4 (the vimwiki ones) are gone
* the vimwiki page from former third tab where I ran the rename command appears in second tab, in the outline pane

**Comment by ranebrown** [comment]
I tested this with a minimal setup and saw the same behavior. Not sure when one of us will have a chance to debug but it is on the list.

**Comment by ranebrown** [comment]
This doesn't seem at all related to VOoM or any other plugins.

**Comment by tinmarino** [comment]
First I thought it was necessary to save windows layout, so I paste the code here for memory.
Heavilly copied from [this answer](https://vi.stackexchange.com/a/22545/5026).


```vim
function vimwiki#u#save_win_layout() abort
  if !(v:version > 801 || has('patch-8.1.0307'))
    return
  endif
  let s:layout = winlayout()
  let s:resize_cmd = winrestcmd()
  call s:add_buf_to_layout(s:layout)
endfunction


function vimwiki#u#restore_win_layout() abort
  if !(v:version > 801 || has('patch-8.1.0307'))
    return
  endif
  " Clause: must be save first
  if !exists('s:layouts')
    return
  endif

  " Create clean window
  new
  wincmd o

  " Restore buffers recursively
  call s:apply_layout(s:layout)

  " Resize windows
  exe s:resize_cmd
endfunction


" Replace buffer in layout
" Warning give me an exetended layout: passed in add_buf
function vimwiki#u#replace_buf_in_layout(buf_old, buf_new, ...) abort
  if !(v:version > 801 || has('patch-8.1.0307'))
    return
  endif
  " Get in
  let layout = a:0 ? a:1 : s:layout

  " Leave: Try to replace if in leaf
  if layout[0] ==# 'leaf'
    if len(layout) > 2 && layout[2] == a:buf_old
      let layout[2] = a:buf_new
    endif
    return
  endif

  " Recurse otherwise
  for child_layout in layout[1]
    call vimwiki#u#replace_buf_in_layout(a:buf_old, a:buf_new, child_layout)
  endfor
endfunction


" Append bufnr to leaves: Recursively parse the layout list
" <- ['row', [['leaf', 1002], ['leaf', 1005]]]
" -> ['row', [['leaf', 1002, 1], ['leaf', 1005, 6]]]
function s:add_buf_to_layout(layout) abort
  " Leave: Append bufnr if in leaf
  if a:layout[0] ==# 'leaf'
    call add(a:layout, winbufnr(a:layout[1]))
    return
  endif

  " Recurse otherwise
  for child_layout in a:layout[1]
    call s:add_buf_to_layout(child_layout)
  endfor
endfunction


" Apply saved layout
" Magical !
function s:apply_layout(layout) abort
  " Leave: Load buffer for leaf
  if a:layout[0] ==# 'leaf'
    if bufexists(a:layout[2])
      exe printf('b %d', a:layout[2])
    endif
    return
  endif

  " Split cols or rows, split n-1 times
  let split_method = a:layout[0] ==# 'col' ? 'rightbelow split' : 'rightbelow vsplit'
  let wins = [win_getid()]
  for child_layout in a:layout[1][1:]
    exe split_method
    let wins += [win_getid()]
  endfor

  " Recurse into child windows
  for index in range(len(wins) )
    call win_gotoid(wins[index])
    call s:apply_layout(a:layout[1][index])
  endfor
endfunction
```


But playing with the test for old vi version without `winlayout()` function, I realized that it was due to the buffer wipeout done in the function, just removing that line solved the issue. Because wipeout and restore destroy the buffers -> windows -> tabs.

Onces again thanks the tests lead the implementation. (It took me 5h to realize that 4 lines to delete)

**Comment by tinmarino** [comment]
Also made a big tests that is really hard to debug => Deleted

```
Do (Testing Keep window layout #592 {{{2):
# Set screen windows
# +-------+--------+
# |       | zzz    |
# |  Comp |        |
# |  ---  |  ----  |
# | new2  |  new1  |
# +-------+--------+
  :Log 'Creating'\<Cr>
  :edit $HOME/testmarkdown/Test-Rename-Layout.md\<Cr>
  :call WriteMe()\<Cr>
  :Log 'Creating 2'\<Cr>
# Left
  :AssertEqual expand('%'), $HOME . '/testmarkdown/Test-Rename-Layout.md'\<Cr>
  :Log 'Creating 3'\<Cr>
# Right
  :exe 'vsplit ' . $HOME . '/testmarkdown/Test-Rename-zzz.md'\<Cr>
  :AssertEqual expand('%:t'), 'Test-Rename-zzz.md'\<Cr>
  :Log 'Creating 4'\<Cr>
  :exe 'split ' . $HOME . '/testmarkdown/new1'\<Cr>
  :AssertEqual expand('%:t'), 'new1'\<Cr>
  \<C-w>k
  :AssertEqual expand('%:t'), 'Test-Rename-zzz.md'\<Cr>
  \<C-w>h
# Left
  :Log 'Creating 5'\<Cr>
  :AssertEqual expand('%:t'), 'Test-Rename-Layout.md'\<Cr>
  :exe 'split ' . $HOME . '/testmarkdown/new2'\<Cr>
  :AssertEqual expand('%:t'), 'new2'\<Cr>
# # Rename main
  :Log 'Renaming'\<Cr>
  \<C-w>h\<C-w>k
  :AssertEqual expand('%:t'), 'Test-Rename-Layout.md'\<Cr>
  :AssertEqual 'vimwiki', &ft\<Cr>
  :VimwikiRenameFile Test-Rename-Layout-new1\<Cr>
  \<C-w>h\<C-w>k
  :AssertEqual expand('%'), $HOME . '/testmarkdown/Test-Rename-Layout-new1.md'\<Cr>
# Check right bottom
  :Log 'Checking 1'\<Cr>
  \<C-w>l\<C-w>j
  :AssertEqual expand('%:t'), 'new1'\<Cr>
# Check right top
  :Log 'Checking 2'\<Cr>
  \<C-w>l\<C-w>k
  :AssertEqual expand('%:t'), 'Test-Rename-zzz.md'\<Cr>
# Check left bottom
  :Log 'Checking 3'\<Cr>
  \<C-w>h\<C-w>j
  :AssertEqual expand('%:t'), 'new2'\<Cr>
# Check left top
  :Log 'Checking 4'\<Cr>
  \<C-w>h\<C-w>k
  :AssertEqual expand('%:t'), 'Test-Rename-Layout-new1.md'\<Cr>
```

**Comment by tinmarino** [comment]
As `win_getid()` is only present for version > 800, the tests are not watching for the windows order / position: they would be horrible to maintain. Plus, they must be implemented with `Do` otherwise they freeze the `run_test.sh` script.

So removing the wipeout

```vim
  " Remove wiki buffers
  for bitem in blist
    execute 'bwipeout '.escape(bitem[0], ' ')
  endfor
```

Got this issue fixed: 8cc99c00c2fa03b
